File: projects/playground/src/proxies/hwui/TextCropper.cpp
#include "TextCropper.h"

Glib::ustring TextCropper::shortenStringIfNeccessary(std::shared_ptr<Font> font, const Glib::ustring &text,
                                                     double maxWidth)
{
  if(font->getStringWidth(text) > maxWidth)
    return preserveEndShorten(font, text, maxWidth);
  return text;
}

Glib::ustring TextCropper::preserveEndShorten(const std::shared_ptr<Font> &font, const Glib::ustring &text,
                                              double maxWidth)
{
  const auto textLen = text.length();
  const auto min = 0;
  auto edgeCaseString = edgeCases(min, textLen, text);
  if(!edgeCaseString.empty())
  {
    return edgeCaseString;
  }

  if(font->getStringWidth("..") > maxWidth)
    return "..";

  auto spaceSplitString = spaceBasedCropping(text, font, maxWidth * 0.95);
  if(!spaceSplitString.empty())
  {
    return spaceSplitString;
  }

  return rigidSplit(font, text, maxWidth);
}

Glib::ustring TextCropper::rigidSplit(const std::shared_ptr<Font> &font, const Glib::ustring &text,
                                      const double maxWidth)
{

  Glib::ustring front;
  Glib::ustring back;
  const Glib::ustring padding = "..";
  int i = 0;
  while(font->getStringWidth(front + padding) < maxWidth / 2)
  {
    front.push_back(text.at(static_cast<Glib::ustring::size_type>(i++)));
  }
  i = static_cast<int>(text.length() - 1);
  while(font->getStringWidth(back + padding) < maxWidth / 2)
  {
    back.insert(back.begin(), text.at(static_cast<Glib::ustring::size_type>(i--)));
  }
  return front + ".." + back;
}

Glib::ustring TextCropper::edgeCases(const int min, const Glib::ustring::size_type len, const Glib::ustring &str)
{
  if(len == 0)
    return "";
  if(len < min)
    return str;
  if((min + 1) == len)
    return str.substr(0, static_cast<Glib::ustring::size_type>(min + 1)) + "..";

  return "";
}

Glib::ustring TextCropper::spaceBasedCropping(const Glib::ustring &text, const std::shared_ptr<Font> &font,
                                              double maxWidth)
{
  const auto &c = text;
  if(maxWidth == 0)
    return "";
  auto roughTwoThirds = (text.length() / 3) * 2;
  if(auto pos = text.find_last_of(' '))
  {
    if(pos < text.length() && pos >= roughTwoThirds)
    {
      auto partAfterSpace = c.substr(pos + 1, c.length() - 1);
      Glib::ustring front = "";
      for(int i = 0; (font->getStringWidth(to_string(front + partAfterSpace + "...")) < maxWidth); i++)
      {
        front += text[i];
      }
      return front + ".." + partAfterSpace;
    }
  }
  return "";
}

File: projects/epc/web-ui/src/main/java/com/nonlinearlabs/client/world/TextCropper.java
package com.nonlinearlabs.client.world;

import com.google.gwt.canvas.dom.client.Context2d;

public class TextCropper {
	public static String crop(String text, Context2d ctx, double maxWidth) {
		int min = 0;
		int max = text.length();

		if (ctx.measureText(text).getWidth() <= maxWidth)
			return text;

		return customAlgorithmAlaJustus(text, min, max, ctx, maxWidth);
	}

	private static String customAlgorithmAlaJustus(String text, int min, int textLen, Context2d ctx, double maxWidth) {
		String edgeCaseString = edgeCases(min, textLen, text);
		if (edgeCaseString != null) {
			return edgeCaseString;
		}

		String spaceSplitString = spaceBasedCropping(text, textLen, ctx, maxWidth * 0.95);
		if (spaceSplitString != null) {
			return spaceSplitString;
		}

		final String padding = "..";
		StringBuilder front = new StringBuilder();
		int i = 0;
		while (ctx.measureText(front.toString() + padding).getWidth() < maxWidth / 2) {
			front.append(text.charAt(i++));
		}
		StringBuilder back = new StringBuilder();
		i = text.length() - 1;
		while (ctx.measureText(back.toString() + padding).getWidth() < maxWidth / 2) {
			back.insert(0, text.charAt(i--));
		}

		String ret = front.toString() + ".." + back.toString();
		return ret;
	}

	private static String edgeCases(int min, int max, String text) {
		if (max == 0)
			return "";
		if (text.length() < min)
			return text;
		if ((min + 1) == max)
			return text.substring(0, min) + "..";
		return null;
	}

	private static String spaceBasedCropping(String text, int max, Context2d ctx, double maxWidth) {

		int lastSpace = text.lastIndexOf(" ");

		if (lastSpace >= (text.length() / 3) * 2) {
			String partAfterSpace = text.substring(lastSpace + 1);

			if (ctx.measureText(partAfterSpace).getWidth() >= maxWidth / 2)
				return null;

			StringBuilder front = new StringBuilder();
			for (int i = 0; ctx.measureText(front + partAfterSpace + "..").getWidth() < maxWidth; i++) {
				front.append(text.charAt(i));
			}
			return front + ".." + partAfterSpace;
		}
		return null;
	}
}
